doc/src/Developer_write_pair.rst

```diff
@@ -0,0 +1,1196 @@
+
+           if (eflag) evdwl = factor_lj * (aexp - bexp - offset[itype][jtype]);
+           if (evflag) ev_tally(i, j, nlocal, newton_pair, evdwl, 0.0, fpair, delx, dely, delz);
+         }
+       }
+     }
+
+If only the global virial is needed and no energy, then calls to
+``ev_tally()`` can be avoided altogether, and the global virial can be
+computed more efficiently from the dot product of the total per-atom
+force vector and the position vector of the corresponding atom,
+:math:`\vec{F}\cdot\vec{r}`.  This has to be done *after* all pair-wise
+forces are computed and *before* the reverse communication to collect
+data from ghost atoms, since the position has to be the position that was
+used to compute the force, i.e. *not* the "local" position if that ghost
+atom is a periodic copy.
+
+.. code-block:: c++
+
+     if (vflag_fdotr) virial_fdotr_compute();
+   }
+
+
+Computing force and energy for a single pair
+""""""""""""""""""""""""""""""""""""""""""""
+
+Certain features in LAMMPS only require computing interactions between
+individual pairs of atoms and the (optional) ``single()`` function is
+needed to support those features (e.g. for tabulation of force and
+energy with :doc:`pair_write <pair_write>`).  This is a repetition of
+the force kernel in the ``compute()`` function, but only for a single
+pair of atoms, where the (squared) distance is provided as a parameter
+(so it may not even be an existing distance between two specific atoms).
+The energy is returned as the return value of the function and the force
+as the `fforce` reference.  Note, that this is, similar to how *fpair*
+is used in the ``compute()`` function, the magnitude of the force along
+the vector between the two atoms *divided* by the distance.
+
+The ``single()`` function is optional.  The member variable
+`single_enable` should be set to 0 in the constructor, if it is not
+implemented (its default value is 1).
+
+.. code-block:: c++
+
+   /* ---------------------------------------------------------------------- */
+
+   double PairBornGauss::single(int /*i*/, int /*j*/, int itype, int jtype, double rsq,
+                                double /*factor_coul*/, double factor_lj, double &fforce)
+   {
+     double r, dr, aexp, bexp;
+
+     r = sqrt(rsq);
+     dr = r - r0[itype][jtype];
+     aexp = biga0[itype][jtype] * exp(-alpha[itype][jtype] * r);
+     bexp = biga1[itype][jtype] * exp(-beta[itype][jtype] * dr * dr);
+
+     fforce = factor_lj * (alpha[itype][jtype] * aexp - 2.0 * dr * beta[itype][jtype] * bexp) / r;
+     return factor_lj * (aexp - bexp - offset[itype][jtype]);
+   }
+
+
+Reading and writing of restart files
+""""""""""""""""""""""""""""""""""""
+
+Support for writing and reading binary restart files is provided by the
+following four functions.  Writing is only done by MPI processor rank 0.
+The output of global (not related to atom types) settings is delegated
+to the ``write_restart_settings()`` function.  Implementing the
```

Comment:
Explain that if the pair style only has a handful of per-type params, it should define these methods.
If there are large tabulation files (e.g. EAM) or manybody it should not.